### PR TITLE
Modified copyTable to use chunked queries

### PR DESCRIPTION
**Pull request by Yetangitu**
Database type conversion can run out of memory on low-memory systems or systems with large _DBPREFIX__filecache tables. An attempted conversion (mysql->postgresql) of an installation with ~12M rows in _DBPREFIX__filecache ran out of memory on a 32-bit system with 2GB RAM/4GB swap once the php process had reached a size of ~1.5GB.

Breaking up the query in chunks alleviates this issue. The default chunk size is 100000, this can be reduced (or increased) by using the --chunk-size option.

While there are other possible solutions to this issue (disabling the query cache being one of them), the presented one does not require any intervention by the user and gives the highest performance.

This patch has only been tested for conversion from mysql/mariadb to postgresql. The chunking is based on 'LIMIT ... OFFSET ...' which is supported by sqlite, mysql/mariadb and postgresql so it 'should work' in any direction, but this has not been tested.

### CODE REVIEW

**Comment by Yetangitu** [comment]
...this contribution is MIT licensed...

**Comment by MorrisJobke** [comment]
cc @DeepDiver1975 @PVince81 @rullzer

**Comment by PVince81** [comment]
@icewind1991

**Comment by PVince81** [comment]
ah, I think @bantu wrote that tool

**Comment by bantu** [comment]
@PVince81 Yes, it was me.

**Comment by bantu** [comment]
@Yetangitu Does LIMIT (with OFFSET) guarantee constant ordering without ORDER BY?

**Comment by Yetangitu** [comment]
The opinions on that differ, but most of what I've seen (for mysql/mariadb) points the other way: constant ordering is not guaranteed when using ORDER BY in combination with LIMIT ... OFFSET vs straight SELECT due to the way the query optimizer handles these queries (https://dev.mysql.com/doc/refman/5.7/en/limit-optimization.html, https://bugs.mysql.com/bug.php?id=72076). In my test case (~12M rows from mariadb (using InnoDB) to postgresql) all data was transfered without overlap. 

As far as I know sqlite returns data in an arbitrary, but deterministic order when doing a straight SELECT. As to how Postgresql reacts I do not have enough information yet.

While it would be possible to add an ORDER BY clause, this is not straightforward (what should the data be ordered by?) and it does affect performance.

Another factor which might affect the way mysql/mariadb reacts in this case is the used storage engine. While I doubt there are many Owncloud installations based on MyISAM, they might exist. As to how MyISAM reacts here I, again, do not have enough information.

**Comment by bantu** [comment]
Probably the following can be used to make the query deterministic: If chunking is disabled, don't use an order clause. Else If the table has a primary key, order by primary key. Else, order by columnA, columnB, ..., column.

**Comment by PVince81** [comment]
@Yetangitu looks like you pushed https://github.com/Yetangitu/core/commit/2ca0c9ab1d5aa9724727e0058c6ddb9bba8e450a on the wrong branch ? It is not related to this PR but I remember it being part of another discussion regarding mime types.

What's the next step for this PR here ?

**Comment by Yetangitu** [comment]
It looks like something went wrong with an unrelated pull request (the migration thing) which for some reason included this change, just ignore this part.

**Comment by MorrisJobke** [comment]
@cmonteroluque I will move this to 8.2 because it's more a feature, than a bug fix.

**Comment by cmonteroluque** [comment]
@MorrisJobke ok

**Comment by rullzer** [comment]
Mmmm this db magic makes me feel somewhat insecure. And I fear that not all db's will behave the same. sqlite->mariadb might react slightly different than mariadb->postgres etc.

If we could have somebody to write integration test we can run from time to time (create an instance with a shitload of file entries for example) and migrate then I'd feel a lot more confortable. Since it does have the potential to eat data.

**Comment by LukasReschke** [comment]
@Yetangitu https://github.com/Yetangitu/core/commit/2ca0c9ab1d5aa9724727e0058c6ddb9bba8e450a seems to be unrelated. Can you please remove this?

Also since it seems that we want integration tests on this before merging let's move this back to 2 instead of 3.

**Comment by Yetangitu** [comment]
Removed the unrelated commit.

**Comment by Yetangitu** [comment]
Looking around the 'net for an answer on the question of result order without using ORDER BY, the consensus seems to be that the only way to get a guaranteed order over time (ie, independent on database state at the time of invocation) is to use an ORDER BY clause. Without ORDER BY, result order can not be guaranteed to be the same due to changes in the database (new records added, VACUUM performed, index changed, etc). But... these events are of no concern for the purpose of this tool, as we can assume that a) the database is not accessed by other clients during migration and b) nobody is going to do database maintenance during migration and c) the indexes are not going to be changed during migration.

Furthermore, even though there is no guaranteed order without ORDER BY, it is not so much any guaranteed order we're after but deterministic behaviour. All we need is for the results to be returned in a deterministic way so that all rows are transferred without duplicates and without losing data.

Looking a bit deeper into the issue of getting a guaranteed deterministic query ordering without knowing any details about the tables to be queried, the easiest solution for sqlite as a source would be to use 'ORDER BY ROWID' (http://www.sqlite.org/lang_createtable.html#rowid), assuming that no tables are declared 'WITHOUT ROWID'. Limited testing shows this to work without any relevant performance penalty.

Neither MySQL/MariaDB or PostgreSQL provide for something similar. PostgreSQL provides the _ctid_ column (http://www.postgresql.org/docs/9.5/static/ddl-system-columns.html) which sounds like it could be used for the same purpose, but actually using it incurs a massive performance penalty. Older PostgreSQL versions also offered an _oid_ column, but this has been disabled by default since v8.1. MySQL/MariaDB does not offer anything resembling ROWID.

So... a deterministic order _can_ be guaranteed for what is likely to be the most common migration case, that from sqlite to MySQL/MariaDB or PostgreSQL. For what is likely to be the second most common migration case, that from MySQL/MariaDB to PostgreSQL this is currently not the case (other than by doing an inspection of the database, using ORDER BY (primary key) for those tables which have one (which could be compound), ORDER BY T.* for those which lack it - at a massive performance penalty). For migrations from PostgreSQL to anything else order can be guaranteed at the price of performance.

Anecdotal evidence (which is worth nothing) shows that the patched version works as intended for the first two migration cases (sqlite->mysql and mysql->postgresql). I have not tested migrations from postresql->anything_else (yet). Without this patch, migrations fail on lower-spec hardware. In other words, _something_ needs to be done for that use case.

**Comment by Yetangitu** [comment]
@owncloud-bot  - look up, all the way on top, and you'll find I already stated this contribution is MIT-licensed: https://github.com/owncloud/core/pull/17548#issuecomment-120172271

**Comment by PVince81** [comment]
Can we move this forward ? @nickvergessen do you know about the order by / limit / offset topic ?

File: apps/files/command/scan.php [hunk 1]
```diff
@@ -72,6 +72,12 @@ protected function configure() {
 				'limit rescan to this path, eg. --path="/alice/files/Music", the user_id is determined by the path and the user_id parameter and --all are ignored'
 			)
 			->addOption(
+				'path_list',
```
Comment (nickvergessen) [inline apps/files/command/scan.php]: Changes in this file belong to a different PR?

File: core/command/db/converttype.php [hunk 2]
```diff
@@ -246,27 +252,36 @@ protected function getTables(Connection $db) {
...
-				$data = array();
-				foreach ($row as $columnName => $value) {
-					$data[$toDB->quoteIdentifier($columnName)] = $value;
-				}
-			}
-			$toDB->insert($table, $data);
-		}
-		$progress->finish();
+                /** @var $progress \Symfony\Component\Console\Helper\ProgressHelper */
+                $chunk_size = $input->getOption('chunk-size') ? $input->getOption('chunk-size') : 100000;
```
Comment (nickvergessen) [inline core/command/db/converttype.php]: Default value should be part of `->addOption(`

File: core/command/upgrade.php [hunk 3]
```diff
@@ -74,6 +74,12 @@ protected function configure() {
 				'skips the database schema migration simulation and update directly'
 			)
 			->addOption(
+				'--skip-mime-repair',
```
Comment (nickvergessen) [inline core/command/upgrade.php]: same here

File: core/command/db/converttype.php [hunk 4]
```diff
@@ -246,27 +252,36 @@ protected function getTables(Connection $db) {
...
+                $chunk_count = ceil($count/$chunk_size);
+                if ($chunk_count > 1) {
+                        $output->writeln('chunked query, ' . $chunk_count . ' chunks');
+                }
+                $progress->start($output, $count);
+                $redraw = $count > $chunk_size ? 100 : ($count > 100 ? 5 : 1);
+                $progress->setRedrawFrequency($redraw);
+                for ($chunk = 0; $chunk < $chunk_count; $chunk++) {
+                        $offset = $chunk * $chunk_size;
+                        $query = 'SELECT * FROM '.$table.' LIMIT '.$chunk_size.' OFFSET '.$offset;
```
Comment (nickvergessen) [inline core/command/db/converttype.php]: I guess we should change this to the getQueryBuilder() then to make sure it keeps working in oracle.

**Comment by nickvergessen** [comment]
Rewrote the patch in https://github.com/owncloud/core/pull/24613

**Comment by nickvergessen** [comment]
Thanks for the idea and the hint anyway @Yetangitu :+1:
